File: services/eval/src/generator/concurrency.ts
/**
 * Run an async mapper over items with at most `limit` concurrent in-flight
 * calls. Preserves input order in the result, regardless of completion order.
 * Failures are surfaced as `null` so the caller can skip individual items
 * without aborting the whole run.
 */
export async function mapWithConcurrency<T, R>(
  items: readonly T[],
  limit: number,
  mapper: (item: T, index: number) => Promise<R>,
): Promise<Array<R | null>> {
  const results: Array<R | null> = new Array(items.length).fill(null);
  let cursor = 0;

  const workers = Array.from({ length: Math.max(1, Math.min(limit, items.length)) }, async () => {
    while (true) {
      const index = cursor++;
      if (index >= items.length) return;
      try {
        results[index] = await mapper(items[index], index);
      } catch {
        results[index] = null;
      }
    }
  });

  await Promise.all(workers);
  return results;
}

File: services/eval/src/rag_eval/generator/concurrency.py
import asyncio
from collections.abc import Awaitable, Callable, Sequence


async def map_with_concurrency[T, R](
    items: Sequence[T],
    limit: int,
    mapper: Callable[[T, int], Awaitable[R]],
) -> list[R | None]:
    """Run an async mapper over items with at most ``limit`` concurrent calls.

    Preserves input order in the result regardless of completion order.
    Failures surface as ``None`` so the caller can skip individual items
    without aborting the whole run.
    """
    if not items:
        return []

    semaphore = asyncio.Semaphore(max(1, min(limit, len(items))))

    async def worker(index: int, item: T) -> R | None:
        async with semaphore:
            try:
                return await mapper(item, index)
            except Exception:
                return None

    return list(await asyncio.gather(*(worker(i, item) for i, item in enumerate(items))))
